Support sourceUri passed on nuget restore/install and dotnet restore or RestoreSources set for msbuild /t:restore command, nuget restore

erdembayar opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Partially Fixes: https://github.com/nuget/home/issues/10948
Fixes: https://github.com/NuGet/Home/issues/10961

Regression? Last working version: n/a

## Description
Users can pass sources with `-source` option to the restore/install command. This PR for adding package namespace filtering with `source` option support.  Following scenarios are addressed here:

```
nuget.exe restore -source
RestoreSources in csproj for PackageReference (msbuild/t:restore, nuget restore)
nuget.exe install -source
dotnet.exe restore -source (Already works, added unit tests)
PMC install-package -source
PMC update-package -source
```

Following is not implemented, because https://github.com/NuGet/Home/issues/11035 is not complete yet, so I created follow up https://github.com/NuGet/Home/issues/11140.
`dotnet add package -source`

Also this PR addresses `nuget install` scenarios not covered by previous PRs(`nuget restore`).
```
nuget install packages.config
nuget install Contoso --Version 1.0.0
```
  
## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

erdembayar committed:
User sourceUri instead of package source key so we can pass source Uri from cli.

erdembayar committed:
Fix tests

erdembayar committed:
Add `-source ../sources` tests

erdembayar committed:
Fix merge

erdembayar committed:
Add test sourceuri with package namespace.

erdembayar committed:
Remove debugger

erdembayar committed:
Fix null handling

erdembayar committed:
Fix sourceuri not used for ResolverGather

erdembayar committed:
Fix -source for PackageReference type.

erdembayar committed:
Fix failing tests

erdembayar committed:
Packages.config project doesn't call RestoreArgs.GetEffectiveSources for sources passed through RestoreSources parameter for msbuild restore.

erdembayar committed:
Fix broken MSBuildRestoreTestBroken by refactoring.

erdembayar committed:
Handle if source resolve to null due to Package namespace section name and source name doesn't match.

erdembayar committed:
Handle if actual source name and package namespace key are different casing.

erdembayar committed:
Fix broken tests.

erdembayar committed:
Fix format

erdembayar committed:
Ignore unmatching source for package namespaces

erdembayar committed:
Fix RestoreSources property inside csproj is not working.

erdembayar committed:
Add more tests

erdembayar committed:
'nuget install -source' tests.

erdembayar committed:
Fix 'nuget install packages.config' is not working.

erdembayar committed:
Fix test

erdembayar committed:
msbuild restore csproj with restoresources property

erdembayar committed:
Fix msbuild restore test leaking nupkg into machine level nuget cache folder.

erdembayar committed:
Add install/update with source option for PMC

erdembayar committed:
Revert unwanted change.

erdembayar committed:
Fix failing test.

erdembayar committed:
Clean up

erdembayar committed:
Add PMC install-package End2end tests.

erdembayar committed:
Improve test names

erdembayar committed:
Add dotnet restore tests.

erdembayar committed:
Move create settings part before creating actual solution just in case.

nkolev92 commented:
* Title says packages.config, but body mentions `RestoreSources in csproj for PackageReference (msbuild/t:restore, nuget restore)` so there's an inconsistency
* dotnet restore -source with transitive dependency, it looks there is a bug with dotnet restore not filtering correctly transitive dependencies. - Can you double check this? Kartheek had some feedback in the teams chat. 

i'm reviewing the rest of the PR rn.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageNamespaces/PackageNamespacesConfiguration.cs:
```diff
@@ -38,11 +50,19 @@ public IReadOnlyList<string> GetConfiguredPackageSources(string term)
         }
 
-        internal PackageNamespacesConfiguration(Dictionary<string, IReadOnlyList<string>> namespaces)
+        internal PackageNamespacesConfiguration(Dictionary<string, IReadOnlyList<string>> namespaces, PackageNamespacesProvider packageNamespacesProvider)
```
I think it's unnecessary to pass down a PackageNamespaceProvider here because we're only using it for the new `GetPackageSources` method. 
It'd be cleaner to provide the list of package sources instead.

nkolev92 replied:
I'd also recommend a null check to have the correct expectations about the API.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageNamespaces/PackageNamespacesProvider.cs:
```diff
@@ -58,5 +58,11 @@ public void AddOrUpdatePackageSourceNamespace(PackageNamespacesSourceItem packag
             _settings.SaveToDisk();
         }
+
+        internal IReadOnlyList<PackageSource> GetPackageSources()
```
See above comment. I think this is unnecessary.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs:
```diff
@@ -24,7 +24,9 @@ internal SearchTree(PackageNamespacesConfiguration configuration)
                 foreach (string namespaceId in namespacePerSource.Value)
                 {
-                    Add(namespacePerSource.Key, namespaceId);
+                    PackageSource packageSource = configuration.GetPackageSource(namespacePerSource.Key);
```
Alternatively, you could consider changing `internal PackageNamespacesConfiguration(Dictionary<string, IReadOnlyList<string>> namespaces)` to something that contains a tuple or package source as the key.

Might make also make it more complicated. 
Something to consider.

erdembayar replied:
I tried it but it looks more complicated, for example `NuGet.Configuration` target net45 which doesn't have ValueTuple and also I have to make many more changes to `SearchTreeTest.cs`. This part is still internal and we can address it later without breaking api if it's really necessary.

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs:
```diff
@@ -1823,6 +1826,209 @@ public async Task InstallCommand_SpecifyUnlistedVersion_InstallUnlistedPackagesA
+
+                var opensourceRepositoryPath = Path.Combine(workingPath, "PublicRepository");
+                Directory.CreateDirectory(opensourceRepositoryPath);
+                Util.CreateTestPackage("测试更新包", "1.0.0", opensourceRepositoryPath);
```
nit: Not sure there's a benefit to adding tests with special characters.

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs:
```diff
@@ -3051,6 +3051,435 @@ public void RestoreCommand_NoPackageNamespacesection_NoSourceRelatedLogMessage()
+        }
+
+        [Fact]
+        public async Task RestoreCommand_NameSpaceFilter_PR_WithAllRestoreSources_Properies_Succeed()
```
It might be better if the PackageReference tests are added RestoreNetTest class with the rest of the PackageReference tests.

erdembayar committed:
Address PR comments

erdembayar committed:
Address PR comment about moving tests.

erdembayar committed:
Revert unneeded change.

erdembayar committed:
Fix failing tests.

erdembayar commented:
@nkolev92 
Thank you for your review.
> * Title says packages.config, but body mentions `RestoreSources in csproj for PackageReference (msbuild/t:restore, nuget restore)` so there's an inconsistency

Please review again. I reworded PR title. 

> * dotnet restore -source with transitive dependency, it looks there is a bug with dotnet restore not filtering correctly transitive dependencies. - Can you double check this? Kartheek had some feedback in the teams chat.
> 
> i'm reviewing the rest of the PR rn.

 I can't repro that scenario again, so removed the comment about it.

kartheekp-ms commented:
I cloned https://github.com/NuGet/Samples/tree/main/PackageNamespacesExample and executed ` dotnet restore -s https://pkgs.dev.azure.com/dnceng/public/_packaging/nuget-build/nuget/v3/index.json` command. The result is `restore` successfully downloaded the packages from `nuget-build` source configured in the `NuGet.Config` file and failed to download packages from `nuget` source because the URL was not passed.

I think `--source` or `RestoreSources` option is automatically supported for `PackageReference` project if values provided are subset or equal to the `package sources and namespaces` defined in the `NuGet.Config` file. AFAIK, no additional code is required to support `--source` option for PR scenarios.

erdembayar commented:
> dotnet restore -s https://pkgs.dev.azure.com/dnceng/public/_packaging/nuget-build/nuget/v3/index.json

Did you clear your global nupkg cache before restore? It failed for me.

```
C:\NuGetProj\test\Samples\PackageNamespacesExample> dotnet restore -s https://pkgs.dev.azure.com/dnceng/public/_packaging/nuget-build/nuget/v3/index.json
  Determining projects to restore...
C:\NuGetProj\test\Samples\PackageNamespacesExample\Lib1\Lib1.csproj : error NU1100: Unable to resolve 'NuGet.Commands (>= 5.10.0)' for '.NETFramework,Version=v4.7.2'. [C:\NuGetProj\test\Samples\PackageNamespacesExample\PackageNamespacesExample.sln]
C:\NuGetProj\test\Samples\PackageNamespacesExample\Lib1\Lib1.csproj : error NU1100: Unable to resolve 'NuGet.Configuration (>= 3.3.0)' for '.NETFramework,Version=v4.7.2'. [C:\NuGetProj\test\Samples\PackageNamespacesExample\PackageNamespacesExample.sln]
```

kartheekp-ms commented:
> Did you clear your global nupkg cache before restore? It failed for me.

Yes, Erick I cleared the cache and restore failed for me too. I updated https://github.com/NuGet/NuGet.Client/pull/4188#issuecomment-899888419 comment to reflect that. restore failed because it can't access `nuget` source as it was not passed as a value for `--source` option. This is expected behavior.

erdembayar commented:
> > Did you clear your global nupkg cache before restore? It failed for me.
> 
> Yes, Erick I cleared the cache and restore failed for me too. I updated [#4188 (comment)](https://github.com/NuGet/NuGet.Client/pull/4188#issuecomment-899888419) comment to reflect that. restore failed because it can't access `nuget` source as it was not passed as a value for `--source` option. This is expected behavior.

Sorry, I'm checking your comment again. I added tests to confirm it's working as expected for `--source` option with PR.
My changes for PR are for logging, currently verbosity log looks 
```
         Package namespace matches found for package ID 'A' are: 'nuget-build'.
         Package namespace matches found for package ID 'NuGet.Commands' are: 'nuget'.
         Package namespace matches found for package ID 'NuGet.Commands' are: 'nuget'.
```

After my change it looks below, users really doesn't care about namespaceId but they care about actual source used in restore:

```
         Package namespace matches found for package ID 'A' are: 'https://pkgs.dev.azure.com/dnceng/public/_packaging/nuget-build/nuget/v3/index.json'.
         Package namespace matches found for package ID 'NuGet.Commands' are: 'https://api.nuget.org/v3/index.json'.
         Package namespace matches found for package ID 'NuGet.Commands' are: 'https://api.nuget.org/v3/index.json'.
```

kartheekp-ms commented:
> After my change it looks below, users really doesn't care about namespaceId but they care about actual source used in restore:
> 
> ```
>          Package namespace matches found for package ID 'A' are: 'https://pkgs.dev.azure.com/dnceng/public/_packaging/nuget-build/nuget/v3/index.json'.
>          Package namespace matches found for package ID 'NuGet.Commands' are: 'https://api.nuget.org/v3/index.json'.
>          Package namespace matches found for package ID 'NuGet.Commands' are: 'https://api.nuget.org/v3/index.json'.
> ```

[NU1101](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1101),  NU1102, NU1103 error codes display `package source key` instead of `package source uri`. Hence I think logging `key` is enough.  If there are no namespaces defined on the `source uri` then restore will fail because namespaces are enabled but effective sources and namespaces configuration don't match. Please let me know if I am missing something.

erdembayar committed:
1

erdembayar commented:
> > After my change it looks below, users really doesn't care about namespaceId but they care about actual source used in restore:
> > ```
> >          Package namespace matches found for package ID 'A' are: 'https://pkgs.dev.azure.com/dnceng/public/_packaging/nuget-build/nuget/v3/index.json'.
> >          Package namespace matches found for package ID 'NuGet.Commands' are: 'https://api.nuget.org/v3/index.json'.
> >          Package namespace matches found for package ID 'NuGet.Commands' are: 'https://api.nuget.org/v3/index.json'.
> > ```
> 
> [NU1101](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1101), NU1102, NU1103 error codes display `package source key` instead of `package source uri`. Hence I think logging `key` is enough. If there are no namespaces defined on the `source uri` then restore will fail because namespaces are enabled but effective sources and namespaces configuration don't match. Please let me know if I am missing something.

Ok. I'll revert dotnet related code changes. Would you like me remove dotnet related unit/integration tests too?

erdembayar committed:
2

erdembayar committed:
Fix failing tests

erdembayar committed:
3

erdembayar committed:
Add back change.

kartheekp-ms commented:
I had an offline discussion with @erdembayar and asked to check if there is any code path in `packages.config` install/restore scenarios where effective sources are identified before restoring packages. Following this approach, we can keep `package sources` and `package namespaces` configuration independent.

https://github.com/NuGet/NuGet.Client/blob/c1a0597e5498bcf3bf0938b3dd0170edbf0b2d20/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs#L146

erdembayar committed:
Adjust tests.

erdembayar committed:
Test

erdembayar committed:
Change from permissive mode to fully specified mode.

erdembayar committed:
Fix install tests.

nkolev92 commented:
Based on our last discussion, I think there's some feedback that needs addressed here. 

@erdembayar can you please ping us when it's ready for review!

Thanks!

erdembayar commented:
@nkolev92 @kartheekp-ms 
I'm closing this PR in favor of https://github.com/NuGet/NuGet.Client/pull/4214/files
It looks all `--source` concern was not an issue, it already works, only thing we needed was unit tests to assert.
